StringIndexOutOfBoundsException in EntityNotDeclaredCodeAction.getEntityName

The following content causes lemminx to throw an exception:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE suite SYSTEM 'https://testng.org/testng-1.0.dtd'>
<sweet
  name='Testos'
  verbose='5'>
  <test name='s'>
    &bar;
    <packages>
      <package name='uitest.*' />
    </packages>
  </test>
</sweet>
``` 


```
[Error - 8:01:16 PM] Sep. 01, 2020 08:01:16 org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError()
Message: Internal error: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
java.util.concurrent.CompletionException: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1309)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1276)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 0
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3756)
	at java.base/java.lang.String.substring(String.java:1902)
	at org.eclipse.lemminx.extensions.contentmodel.participants.codeactions.EntityNotDeclaredCodeAction.getEntityName(EntityNotDeclaredCodeAction.java:229)
	at org.eclipse.lemminx.extensions.contentmodel.participants.codeactions.EntityNotDeclaredCodeAction.doCodeAction(EntityNotDeclaredCodeAction.java:49)
	at org.eclipse.lemminx.extensions.contentmodel.participants.ContentModelCodeActionParticipant.doCodeAction(ContentModelCodeActionParticipant.java:49)
	at org.eclipse.lemminx.services.XMLCodeActions.doCodeActions(XMLCodeActions.java:45)
	at org.eclipse.lemminx.services.XMLLanguageService.doCodeActions(XMLLanguageService.java:255)
	at org.eclipse.lemminx.XMLTextDocumentService.lambda$codeAction$19(XMLTextDocumentService.java:389)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1307)
	... 7 more
```

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java
@@ -46,19 +46,19 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 			SharedSettings sharedSettings, IComponentProvider componentProvider) {
 
 		try {
-			String entityName = getEntityName(diagnostic, range, document);
+			String entityName = getEntityName(diagnostic, document);
 
 			if (entityName == null) {
 				return;
 			}
-			
+
 			DOMDocumentType docType = document.getDoctype();
 			if (docType != null) {
 
 				// Case 1: <!DOCTYPE exists
 				// Add <!ENTITY nbsp "entity-value"> in the subset.
 				// If the subset does not exist, add subset too
-				
+
 				// ie:
 				// <!DOCTYPE article [
 				// <!ELEMENT article (#PCDATA)>
@@ -69,7 +69,7 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 				// Case 2: <!DOCTYPE does not exist
 				// Generate:
 				// <!DOCTYPE article [
-				//     <!ENTITY nbsp "entity-value">
+				// <!ENTITY nbsp "entity-value">
 				// ]>
 				addDoctypeAndEntityCodeAction(entityName, diagnostic, document, sharedSettings, codeActions);
 			}
@@ -79,11 +79,10 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 	}
 
 	/**
-	 * Add a code action that inserts entity declaration to the
-	 * current DOCTYPE's internal subset
+	 * Add a code action that inserts entity declaration to the current DOCTYPE's
+	 * internal subset
 	 * 
-	 * If the internal subset does not exist, the code action
-	 * inserts it as well
+	 * If the internal subset does not exist, the code action inserts it as well
 	 * 
 	 * @param entityName  the entity name for the entity declaration
 	 * @param diagnostic  the code action's diagnostic
@@ -113,7 +112,7 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 			} else {
 				insertString.startUnindentedDoctypeInternalSubset();
 			}
-			
+
 			if (insertPosition.getLine() > 0) {
 				insertString.linefeed();
 			}
@@ -136,8 +135,8 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 	}
 
 	/**
-	 * Add a code action that inserts the DOCTYPE declaration containing
-	 * an internal subset with an entity declaration
+	 * Add a code action that inserts the DOCTYPE declaration containing an internal
+	 * subset with an entity declaration
 	 * 
 	 * @param entityName  the entity name for the entity declaration
 	 * @param diagnostic  the code action's diagnostic
@@ -161,14 +160,12 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 		if (insertPosition.getCharacter() > 0) {
 			insertString.linefeed();
 		}
-		insertString.startDoctype().addParameter(root.getTagName())
-				.startDoctypeInternalSubset().linefeed().indent(1);
+		insertString.startDoctype().addParameter(root.getTagName()).startDoctypeInternalSubset().linefeed().indent(1);
 
 		addEntityDeclaration(entityName, insertString);
 
-		insertString.linefeed()
-				.endDoctypeInternalSubset().closeStartElement();
-		
+		insertString.linefeed().endDoctypeInternalSubset().closeStartElement();
+
 		Position rootStartPosition = document.positionAt(root.getStart());
 		if (insertPosition.getLine() == rootStartPosition.getLine()) {
 			insertString.linefeed();
@@ -219,12 +216,12 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 	 * https://github.com/microsoft/language-server-protocol/issues/887
 	 * 
 	 * @param diagnostic the diagnostic
-	 * @param range      the error range
 	 * @param doc        the DOM document
 	 * @return entity name from the error range and error message.
 	 * @throws BadLocationException
 	 */
-	private String getEntityName(Diagnostic diagnostic, Range range, DOMDocument doc) throws BadLocationException {
+	private static String getEntityName(Diagnostic diagnostic, DOMDocument doc) throws BadLocationException {
+		Range range = diagnostic.getRange();
 		String name = doc.getText().substring(doc.offsetAt(range.getStart()), doc.offsetAt(range.getEnd()));
 		String removedAmpAndSemiColon = name.substring(1, name.length() - 1);
 		if (diagnostic.getMessage().indexOf("\"" + removedAmpAndSemiColon + "\"") < 0) {
@@ -233,8 +230,10 @@ public class EntityNotDeclaredCodeAction implements ICodeActionParticipant {
 		return removedAmpAndSemiColon;
 	}
 
-	private void addEntityDeclaration(String entityName, XMLBuilder builder) {
-		builder.addDeclTagStart("ENTITY").addParameter(entityName)
-				.addParameter("\"entity-value\"").closeStartElement();
+	private static void addEntityDeclaration(String entityName, XMLBuilder builder) {
+		builder.addDeclTagStart("ENTITY") //
+				.addParameter(entityName) //
+				.addParameter("\"entity-value\"")//
+				.closeStartElement();
 	}
 }

--- a/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java
+++ b/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java
@@ -425,6 +425,28 @@ public class DTDDiagnosticsTest {
 	}
 
 	@Test
+	public void Issue862() throws Exception {
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article\n" + //
+				">\n" + //
+				"<article>\n" + //
+				"	&nbsp;\n" + //
+				"</article>";
+
+		Diagnostic d = d(4, 1, 4, 7, DTDErrorCode.EntityNotDeclared,
+				"The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+
+		xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE article\n" + //
+				">\n" + //
+				"<article>\n" + //
+				"	&|nbsp;\n" + // set the range
+				"</article>";
+		testCodeActionsFor(xml, d, ca(d, te(2, 0, 2, 0, "[\n\t<!ENTITY nbsp \"entity-value\">\n]")));
+	}
+	
+	@Test
 	public void EntityNotDeclaredDoctypeEmptySubset() throws Exception {
 		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
 				"<!DOCTYPE article []>\n" + //

